llm/server.go

```diff
@@ -0,0 +1,807 @@
+			s.Close()
+			finalErr = err
+			continue
+		}
+		return s, nil
+	}
+
+	slog.Error("unable to load any llama server", "error", finalErr)
+	return nil, finalErr
+}
+
+type ServerStatus int
+
+const ( // iota is reset to 0
+	ServerStatusReady ServerStatus = iota
+	ServerStatusNoSlotsAvaialble
+	ServerStatusLoadingModel
+	ServerStatusNotResponding
+	ServerStatusError
+)
+
+type ServerStatusResp struct {
+	Status          string `json:"status"`
+	SlotsIdle       int    `json:"slots_idle"`
+	SlotsProcessing int    `json:"slots_processing"`
+	Error           string `json:"error"`
+}
+
+func (s *LlamaServer) getServerStatus(ctx context.Context) (ServerStatus, error) {
+	// Fail fast if its exited
+	if s.cmd.ProcessState != nil {
+		return ServerStatusError, fmt.Errorf("llama runner process no longer running: %d", s.cmd.ProcessState.ExitCode())
+	}
+
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, fmt.Sprintf("http://127.0.0.1:%d/health", s.port), nil)
+	if err != nil {
+		return ServerStatusError, fmt.Errorf("error creating GET request: %v", err)
+	}
+	req.Header.Set("Content-Type", "application/json")
+
+	resp, err := http.DefaultClient.Do(req)
+	if err != nil {
+		if errors.Is(err, context.DeadlineExceeded) {
+			return ServerStatusNotResponding, fmt.Errorf("server not responding")
+		}
+		return ServerStatusError, fmt.Errorf("health resp: %w", err)
+	}
+	defer resp.Body.Close()
+
+	body, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return ServerStatusError, fmt.Errorf("read health request: %w", err)
+	}
+
+	var status ServerStatusResp
+	if err := json.Unmarshal(body, &status); err != nil {
+		return ServerStatusError, fmt.Errorf("health unmarshal encode response: %w", err)
+	}
+
+	switch status.Status {
+	case "ok":
+		return ServerStatusReady, nil
+	case "no slot available":
+		return ServerStatusNoSlotsAvaialble, nil
+	case "loading model":
+		return ServerStatusLoadingModel, nil
+	default:
+		return ServerStatusError, fmt.Errorf("server error: %+v", status)
+	}
+}
+
+func (s *LlamaServer) Ping(ctx context.Context) error {
+	_, err := s.getServerStatus(ctx)
+	if err != nil {
+		slog.Debug("server unhealthy", "error", err)
+		return err
+	}
+	return nil
+}
+
+func (s *LlamaServer) waitUntilRunning() error {
+	start := time.Now()
+	expiresAt := time.Now().Add(3 * time.Minute) // be generous with timeout, large models can take a while to load
+	ticker := time.NewTicker(200 * time.Millisecond)
+	defer ticker.Stop()
+
+	slog.Info("waiting for llama runner to start responding")
+	var lastStatus ServerStatus = -1
+	for {
+		select {
+		case err := <-s.done:
+			return fmt.Errorf("llama runner process has terminated: %v", err)
+		case <-ticker.C:
+			if time.Now().After(expiresAt) {
+				// timeout
+				return fmt.Errorf("timed out waiting for llama runner to start")
+			}
+			if s.cmd.ProcessState != nil {
+				return fmt.Errorf("llama runner process no longer running: %d", s.cmd.ProcessState.ExitCode())
```

Comment:
This error will likely be returned for many different "root" errors, which will cause many people to end up in issues not related to their actual problems.

You can see in a previous implementation that I redirected the stderr to a function that intercepts the log messages to track the actual error. If possible I'd like to do something similar again.
https://github.com/ollama/ollama/blob/5e7fd6906f4653fa671aa5d2e2d4dd5bdf17fd36/llm/llama.go#L430

Reply:
Thanks! I'll port this code over.

Reply:
With the current setup, this is going to have somewhat limited value.  With the current structure, the common scenario is some failure in the GPU runner, and the retry logic means we'll switch to the CPU runner(s) and the captured last error will only be reported in the log.  If there's a problem with the model, then we'd iterate through, all the runners would fail, and we would bubble that back up through the API.

In the future it may be interesting to explore warnings bubbling up in the API so that we could capture this fallback scenario and report why we couldn't load on the GPU but are running on CPU.